python-frontend/typeshed_serializer/serializer/symbols.py

```diff
@@ -0,0 +1,358 @@
+        pb_parameter = symbols_pb2.ParameterSymbol()
+        pb_parameter.name = self.name
+        pb_parameter.kind = symbols_pb2.ParameterKind.Value(self.kind.name)
+        pb_parameter.has_default = self.has_default
+        if self.type_annotation is not None:
+            pb_parameter.type_annotation.CopyFrom(self.type_annotation.to_proto())
+        return pb_parameter
+
+
+class OverloadedFunctionSymbol:
+    def __init__(self, overloaded_func_def: mpn.OverloadedFuncDef):
+        self.name = overloaded_func_def.name
+        self.fullname = overloaded_func_def.fullname
+        self.is_final = overloaded_func_def.is_final
+        self.is_property = overloaded_func_def.is_property
+        self.is_static = overloaded_func_def.is_static
+        self.definitions = []
+        for item in overloaded_func_def.items:
+            if isinstance(item, mpn.FuncDef):
+                # Should not happen?
+                self.definitions.append(FunctionSymbol(item))
+            if isinstance(item, mpn.Decorator):
+                self.definitions.append(FunctionSymbol(item.func, decorators=item.decorators))
+
+    def to_proto(self) -> symbols_pb2.OverloadedFunctionSymbol:
+        pb_overloaded_func = symbols_pb2.OverloadedFunctionSymbol()
+        pb_overloaded_func.name = self.name
+        pb_overloaded_func.fullname = self.fullname
+        pb_overloaded_func.is_final = self.is_final
+        pb_overloaded_func.is_property = self.is_property
+        pb_overloaded_func.is_static = self.is_static
+        for definition in self.definitions:
+            pb_overloaded_func.definitions.append(definition.to_proto())
+        return pb_overloaded_func
+
+
+class FunctionSymbol:
+    def __init__(self, func_def: mpn.FuncDef, decorators=None):
+        self.name = func_def.name
+        self.fullname = func_def.fullname
+        self.return_type = extract_return_type(func_def)
+        self.arguments = extract_arguments(func_def)
+        self.has_decorators = func_def.is_decorated
+        self.is_abstract = func_def.is_abstract
+        self.is_async_generator = func_def.is_async_generator
+        self.is_awaitable_coroutine = func_def.is_awaitable_coroutine
+        self.is_coroutine = func_def.is_coroutine
+        self.is_final = func_def.is_final
+        self.is_overload = func_def.is_overload
+        self.is_property = func_def.is_property
+        self.is_static = func_def.is_static
+        self.is_class_method = func_def.is_class
+        self.resolved_decorator_names = []
+        if self.has_decorators and decorators is not None:
+            for dec in decorators:
+                if isinstance(dec, mpn.NameExpr):
+                    # decorator full name might not be the actual fully qualified name if it could not be resolved
+                    # TODO: handle "None" case and check for fallbacks
+                    if dec.fullname is not None:
+                        self.resolved_decorator_names.append(dec.fullname)
```

Comment:
I'm changing the `decorators` for `original_decorators` which should keep `classmethod` and `staticmethod` decorators available.
With regard to storing the FQN, I created a new method to handle qualified expressions as well, which was missing.
I believe this should be enough to resolve what we could resolve before.

Reply:
I'm not sure I understand correctly: are you going to store only FQNs for decorators? In that case it might be tricky to extract the "decorator name" info we had in sonar-python